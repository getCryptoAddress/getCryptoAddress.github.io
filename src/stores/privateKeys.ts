import { defineStore } from "pinia";
import { readonly, ref } from "vue";
import generatePrivateKey from "@/libs/PrivateKeys/generatePrivateKey";
import isValidPrivateKeys from "@/libs/PrivateKeys/isValidPrivateKeys";
import type PrivateKeyFormatted from "@/libs/PrivateKeys/types/PrivateKeyFormatted";
import getFormattedPrivateKey from "@/libs/PrivateKeys/getFormattedPrivateKey";

export const usePrivateKeysStore = defineStore("privateKeys", () => {
  const privateKeys = ref<Uint8Array[]>([]);
  const formattedPrivateKeys = ref<PrivateKeyFormatted[]>();

  async function _generatePrivateKeys() {
    formattedPrivateKeys.value = await Promise.all(
      privateKeys.value.map((key) => getFormattedPrivateKey(key))
    );
  }

  async function setPrivateKeys(keys: (number[] | Uint8Array)[]) {
    const copyKeys = keys.map((key) => new Uint8Array([...key]));

    if (!(await isValidPrivateKeys(copyKeys))) {
      throw new Error("Not valid private keys");
    }
    privateKeys.value = copyKeys;
    await _generatePrivateKeys();
  }

  async function generatePrivatesKeys(count: number) {
    const keys = [];
    for (let i = 0; i < count; i += 1) {
      keys.push(await generatePrivateKey());
    }
    privateKeys.value = keys;
    await _generatePrivateKeys();
  }

  return {
    privateKeys: readonly(privateKeys),
    formattedPrivateKeys: readonly(formattedPrivateKeys),
    generatePrivatesKeys,
    setPrivateKeys,
  };
});
