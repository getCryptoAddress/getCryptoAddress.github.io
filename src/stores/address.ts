import { defineStore } from "pinia";
import { computed, readonly } from "vue";
import { usePrivateKeysStore } from "@/stores/privateKeys";
import getAddresses from "@/libs/Address/getAddresses";

export const useAddressStore = defineStore("address", () => {
  const privateKeysStore = usePrivateKeysStore();

  const address = computed(() =>
    privateKeysStore.privateKeys.reduce((result, key) => {
      result.set(key, getAddresses(key, false));
      return result;
    }, new Map())
  );
  const testAddress = computed(() =>
    privateKeysStore.privateKeys.reduce((result, key) => {
      result.set(key, getAddresses(key, true));
      return result;
    }, new Map())
  );

  return {
    address: readonly(address),
    testAddress: readonly(testAddress),
  };
});
