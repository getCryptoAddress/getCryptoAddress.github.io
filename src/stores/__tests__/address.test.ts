import { describe, it, expect, beforeEach } from "vitest";
import { setActivePinia, createPinia } from "pinia";
import { usePrivateKeysStore } from "../privateKeys";
import { useAddressStore } from "../address";
import { privateKey1, privateKey2, privateKey3 } from "./mocks/keys.mock";

describe("[Store] address", () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it("address generated", async () => {
    const privateKeysStore = usePrivateKeysStore();
    const addressStore = useAddressStore();
    const customKeys = [privateKey1, privateKey2, privateKey3];
    await privateKeysStore.setPrivateKeys(customKeys);

    expect(addressStore.address).toMatchSnapshot();
    expect(addressStore.testAddress).toMatchSnapshot();
  });
});
