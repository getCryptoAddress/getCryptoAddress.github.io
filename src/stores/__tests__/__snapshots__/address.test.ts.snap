// Vitest Snapshot v1

exports[`[Store] address > address generated 1`] = `
Map {
  Uint8Array [
    130,
    205,
    117,
    44,
    180,
    113,
    245,
    200,
    213,
    19,
    97,
    187,
    70,
    157,
    86,
    111,
    19,
    131,
    14,
    18,
    196,
    57,
    9,
    203,
    64,
    196,
    19,
    26,
    73,
    35,
    126,
    80,
  ] => {
    "pkh": "19PBzEVhSpaVS1rRGf8cueNN6fLc4Xvohp",
    "tr": "bc1pshthlwxy627gftrakrrg0k48ewagaee8tc6a7qccdp4nfpczhs3sett0f9",
    "wpkh": "bc1qt0e4gfecx8rcldf2udul9j3m603jsjg8hp56j7",
  },
  Uint8Array [
    201,
    59,
    114,
    215,
    208,
    85,
    88,
    95,
    8,
    25,
    187,
    101,
    2,
    64,
    233,
    47,
    162,
    189,
    25,
    66,
    185,
    4,
    61,
    245,
    90,
    86,
    229,
    154,
    136,
    217,
    111,
    133,
  ] => {
    "pkh": "1PovqogNPxWQkzqQtLo5mVcoZfgNN2GMAr",
    "tr": "bc1ppj3ee474242hnla22c9s5gjshfgs7dgyhkazennlf5xuxnndmsespjv2g2",
    "wpkh": "bc1qlged3xzeyz4cqyaxah85dvrd5wt2andx72ym2p",
  },
  Uint8Array [
    2,
    241,
    8,
    120,
    165,
    66,
    56,
    141,
    45,
    212,
    113,
    234,
    18,
    6,
    90,
    33,
    190,
    120,
    117,
    15,
    251,
    178,
    238,
    44,
    141,
    220,
    116,
    78,
    115,
    152,
    67,
    36,
  ] => {
    "pkh": "15YsHzhewCsugaVUpYEX5DT5iKBLS6Sf1Y",
    "tr": "bc1pmu4gyprm60lawlguhl3z0pxenv8wyxwtna2nqshgfx805xslng5s2demee",
    "wpkh": "bc1qx8nkzexkx62zq86gxcm55tys56qw8jrzxhmtsp",
  },
}
`;

exports[`[Store] address > address generated 2`] = `
Map {
  Uint8Array [
    130,
    205,
    117,
    44,
    180,
    113,
    245,
    200,
    213,
    19,
    97,
    187,
    70,
    157,
    86,
    111,
    19,
    131,
    14,
    18,
    196,
    57,
    9,
    203,
    64,
    196,
    19,
    26,
    73,
    35,
    126,
    80,
  ] => {
    "pkh": "mou9HHagFr1kD8L2zE6zjZagxewJvjbNgy",
    "tr": "tb1pshthlwxy627gftrakrrg0k48ewagaee8tc6a7qccdp4nfpczhs3swraqn2",
    "wpkh": "tb1qt0e4gfecx8rcldf2udul9j3m603jsjg8a80ffd",
  },
  Uint8Array [
    201,
    59,
    114,
    215,
    208,
    85,
    88,
    95,
    8,
    25,
    187,
    101,
    2,
    64,
    233,
    47,
    162,
    189,
    25,
    66,
    185,
    4,
    61,
    245,
    90,
    86,
    229,
    154,
    136,
    217,
    111,
    133,
  ] => {
    "pkh": "n4Kt8rmMCywfY7K2bumTbQq8RfH5BLLUQP",
    "tr": "tb1ppj3ee474242hnla22c9s5gjshfgs7dgyhkazennlf5xuxnndmsesk669j9",
    "wpkh": "tb1qlged3xzeyz4cqyaxah85dvrd5wt2andx5vlg3j",
  },
  Uint8Array [
    2,
    241,
    8,
    120,
    165,
    66,
    56,
    141,
    45,
    212,
    113,
    234,
    18,
    6,
    90,
    33,
    190,
    120,
    117,
    15,
    251,
    178,
    238,
    44,
    141,
    220,
    116,
    78,
    115,
    152,
    67,
    36,
  ] => {
    "pkh": "mk4pb3ndkEKATgy6Y7Ctu8fQaJn3Pstsis",
    "tr": "tb1pmu4gyprm60lawlguhl3z0pxenv8wyxwtna2nqshgfx805xslng5sa905rk",
    "wpkh": "tb1qx8nkzexkx62zq86gxcm55tys56qw8jrzv3qctj",
  },
}
`;
